server/src/test/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocatorInvalidWeightsTests.java

```diff
@@ -0,0 +1,303 @@
+            final var allocator = new BalancedShardsAllocator(
+                BalancerSettings.DEFAULT,
+                WriteLoadForecaster.DEFAULT,
+                balancingWeightsFactory
+            );
+
+            final ClusterState clusterState = makeUnbalanced(
+                ClusterStateCreationUtils.state(3, new String[] { "one", "two", "three" }, 1),
+                "node_0"
+            );
+            final RoutingAllocation allocation = new RoutingAllocation(
+                new AllocationDeciders(List.of()),
+                clusterState.getRoutingNodes().mutableCopy(),
+                clusterState,
+                ClusterInfo.EMPTY,
+                null,
+                System.nanoTime()
+            );
+            balancingWeightsFactory.returnInvalidWeightsForRandomNodes(clusterState);
+            assertInvalidWeightsMessageLogged(() -> allocator.allocate(allocation));
+
+            // There should have been no changes
+            assertFalse(allocation.routingNodesChanged());
+        }
+    }
+
+    public void testNodesBalancingIsAbortedWhenWeFirstSeeAnInvalidWeight() {
+        try (var ignored = BalancedShardsAllocator.disableInvalidWeightsAssertionsAndRemoveLogRateLimiting()) {
+            final var balancingWeightsFactory = new InvalidWeightsBalancingWeightsFactory();
+            final var allocator = new BalancedShardsAllocator(
+                BalancerSettings.DEFAULT,
+                WriteLoadForecaster.DEFAULT,
+                balancingWeightsFactory
+            );
+
+            final AllocationDecider allocationDecider = spy(AllocationDecider.class);
+            final ClusterState clusterState = makeUnbalanced(
+                ClusterStateCreationUtils.state(3, new String[] { "one", "two", "three" }, randomIntBetween(1, 2)),
+                "node_0"
+            );
+            final RoutingAllocation allocation = new RoutingAllocation(
+                new AllocationDeciders(List.of(allocationDecider)),
+                clusterState.getRoutingNodes().mutableCopy(),
+                clusterState,
+                ClusterInfo.EMPTY,
+                null,
+                System.nanoTime()
+            );
+
+            doAnswer(iom -> {
+                if (allocation.routingNodes().getRelocatingShardCount() > 0) {
+                    return randomFrom(Float.POSITIVE_INFINITY, Float.NEGATIVE_INFINITY, Float.NaN);
+                }
+                return iom.callRealMethod();
+            }).when(balancingWeightsFactory.getWeightFunction()).calculateNodeWeightWithIndex(any(), any(), any());
+            assertInvalidWeightsMessageLogged(() -> allocator.allocate(allocation));
+            // We should have aborted balancing after that first move
+            assertEquals(1, allocation.routingNodes().getRelocatingShardCount());
+        }
+    }
+
+    public void testMoveShardsIsSkippedWhenInvalidWeightsAreEncounteredDuringSorting() {
```

Comment:
I could see the argument that this is not what we want? perhaps a better strategy here is just to assume any invalid weight is `Float.MAX_VALUE` so that e.g. shutting down nodes can still evacuate to whichever nodes are present? But also hesitant to add too much behaviour here.